go-controller/pkg/ovn/egressip.go

```diff
@@ -2045,28 +2038,102 @@ func (oc *DefaultNetworkController) initEgressIPAllocator(node *kapi.Node) (err
+	var isOldReady, isNewReady, isNewReachable bool
+	var nodeName string
+	if oldNode != nil {
+		oldLabels = oldNode.GetLabels()
+		isOldReady = oc.isEgressNodeReady(oldNode)
+		nodeName = oldNode.Name
+	}
+	if newNode != nil {
+		// Initialize the allocator on every update,
+		// ovnkube-node/cloud-network-config-controller will make sure to
+		// annotate the node with the egressIPConfig, but that might have
+		// happened after we processed the ADD for that object, hence keep
+		// retrying for all UPDATEs.
+		if err := oc.initEgressIPAllocator(newNode); err != nil {
+			klog.Warningf("Egress node initialization error: %v", err)
+		}
+
+		newLabels = newNode.GetLabels()
+		isNewReady = oc.isEgressNodeReady(newNode)
+		isNewReachable = oc.isEgressNodeReachable(newNode)
+		nodeName = newNode.Name
+	} else if oldNode != nil {
+		err := oc.deleteNodeForEgress(oldNode)
+		if err != nil {
+			return nil
+		}
+	}
+
+	_, oldHadEgressLabel := oldLabels[nodeEgressLabel]
+	_, newHasEgressLabel := newLabels[nodeEgressLabel]
+	oc.setNodeEgressAssignable(nodeName, newHasEgressLabel)
+	oc.setNodeEgressReady(nodeName, isNewReady)
+
+	// If the node is not labeled for egress assignment, just return
+	// directly, we don't really need to set the ready / reachable
+	// status on this node if the user doesn't care about using it.
+	if !oldHadEgressLabel && !newHasEgressLabel {
+		return nil
+	}
+
+	if oldHadEgressLabel && !newHasEgressLabel {
+		klog.Infof("Node: %s has been un-labeled, deleting it from egress assignment", nodeName)
+		return oc.deleteEgressNode(nodeName)
+	}
+
+	if !oldHadEgressLabel && newHasEgressLabel {
+		klog.Infof("Node: %s has been labeled, adding it for egress assignment", nodeName)
+		if isNewReady && isNewReachable {
+			oc.setNodeEgressReachable(nodeName, isNewReachable)
+			if err := oc.addEgressNode(nodeName); err != nil {
+				return err
+			}
+		} else {
+			klog.Warningf("Node: %s has been labeled, but node is not ready"+
+				" and reachable, cannot use it for egress assignment", nodeName)
+		}
+		return nil
 	}
-	oc.addEgressNodeFailed.Delete(node.Name)
-	if err := oc.initEgressIPAllocator(node); err != nil {
-		klog.V(5).Infof("Egress node initialization error: %v", err)
+
+	if isOldReady == isNewReady {
+		return nil
+	}
+
+	if !isNewReady {
+		klog.Warningf("Node: %s is not ready, deleting it from egress assignment", nodeName)
+		if err := oc.deleteEgressNode(nodeName); err != nil {
+			return err
+		}
+	} else if isNewReady && isNewReachable {
+		klog.Infof("Node: %s is ready and reachable, adding it for egress assignment", nodeName)
+		oc.setNodeEgressReachable(nodeName, isNewReachable)
+		if err := oc.addEgressNode(nodeName); err != nil {
+			return err
+		}
 	}
+
 	return nil
 }
 
 // deleteNodeForEgress remove the default allow logical router policies for the
 // node and removes the node from the allocator cache.
 func (oc *DefaultNetworkController) deleteNodeForEgress(node *v1.Node) error {
```

Comment:
Please update the comment.
With your changes `deleteNodeForEgress` only removes the node from the allocator cache. Could we rename it to reflect that? It is confusing since we also have `deleteEgressNode`

Reply:
I suggested a new name in the original comment. Let me know what you think.

Reply:
sounds good!